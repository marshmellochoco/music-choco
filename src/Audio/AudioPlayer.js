import {
    useState,
    useEffect
} from "react";

export default function AudioPlayer() {
    const [duration, setDuration] = useState()
    const [curTime, setCurTime] = useState()
    const [volume, setVolume] = useState()
    const [clickedTime, setClickedTime] = useState()
    const [playing, setPlaying] = useState(false)

    useEffect(() => {
        const audio = document.getElementById('audio-player')

        const setAudioData = () => {
            setDuration(audio.duration)
            setCurTime(audio.currentTime)
            setVolume(audio.volume)
        }

        const setAudioTime = () => setCurTime(audio.currentTime)
        const setAudioVolume = () => setVolume(audio.volume)

        audio.addEventListener("loadeddata", setAudioData)
        audio.addEventListener("timeupdate", setAudioTime)
        audio.addEventListener("volumechange", setAudioVolume)

        playing ? audio.play() : audio.pause()

        if (clickedTime && clickedTime !== curTime) {
            audio.currentTime = clickedTime
            setClickedTime(null);
        }

        if (volume && volume !== audio.volume)
            audio.volume = volume; 

        return () => {
            audio.removeEventListener("loadeddata", setAudioData)
            audio.removeEventListener("timeupdate", setAudioTime)
            audio.removeEventListener("volumechange", setAudioVolume)
        }
    }, [playing, clickedTime, curTime, volume])

    return {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    }

}
