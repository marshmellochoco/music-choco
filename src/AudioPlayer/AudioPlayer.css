.player{
    font-size: 0.8em;
    width: 98%;
    padding: 10px 1%;
    background-color: #303030;
    color: #d0d0d0;
    overflow: hidden;
    position: absolute;
    bottom: 0;
}

.player-container {
    display: grid;
    grid-template-columns: 1fr 2fr 1fr;
    grid-template-rows: 2fr 1fr;
}

.player-title{
    margin: auto 0;
    grid-row: 1/3;
    text-align: left;
    display: grid;
    grid-template-columns: 64px 1fr;
    grid-template-rows: 1fr 1fr;
}

.player-title .icon{
    grid-row: 1/3;
    width: 64px;
    height: 64px;
}

.player-title .icon img{
    max-width: 64px;
    max-height: 64px;
    background-color: white;
}

.player-title .title{
    margin: 0 10px;
    display: flex;
    align-items: flex-end;
}

.player-title .artist{
    margin: 0 10px;
    display: flex;
    align-items: flex-start;
}


.player-timeline{
    margin: auto 0;
    text-align: center;
    display: grid;
    grid-template-columns: 1fr 5fr 1fr;
}

.player-timeline .progress{
    overflow: hidden;
    margin: auto 0;
    background-color: grey;
    height: 0.3em;
    border-radius: 2px;
    cursor: pointer;
}

.player-timeline .player-time{
    padding: 0 10%;
    font-size: 0.9em;
}

/* 
.player-timeline input[type=range]{
    overflow: hidden;
    width: 70%;
    -webkit-appearance: none;
    background-color: #303030;
}

input[type=range]::-webkit-slider-runnable-track{
    border: none;
    height: 0.35em;
    background: grey;
    border-radius: 10px;
    cursor: pointer;
}

input[type=range]::-webkit-slider-thumb{
    -webkit-appearance: none;
    height: 0.5em;
    width: 0.5em;
    border-radius: 50%;
    background: white;
    margin-top: -1px;
    cursor: pointer;
    box-shadow: -80px 0 0 80px green;
}

input[type=range]::-webkit-slider-thumb:hover{
    height: 0.8em;
    width: 0.8em;
    border-radius: 50%;
    background: white;
    margin-top: -3px;
}

input[type=range]:focus {
    outline: none;
} */

.player-control{
    text-align: center;
}

.player-control .control-button{
    height: 100%;
    margin: auto;
    width: 1.8em;
    padding: 0px 12px;
}

.player-control .play{
    width: 2.5em;
}

.player-control .control-button:hover{
    color: grey;
}

.player-volume{
    display: flex;
    margin-left: auto;
    width: 30%;
    grid-row: 1/3;
    grid-column: 3;
}

@media only screen and (max-width: 1024px){
    .player-volume input{
        width: 50%;
    }
}

@media only screen and (max-width: 862px){
    input[type=range]{
        width: 60%;
    }
}

@media only screen and (max-width: 480px){
    .player-time{
        display: none;
    }
    .player-timeline input[type=range]{
        width: 100%;
    }
    .player-control .control-button{
        font-size: 1em
    }
}

.slider{
    overflow: hidden;
    margin: auto 0;
    background-color: grey;
    height: 0.3em;
    border-radius: 2px;
    cursor: pointer;
}

.slider input[type=range]{
    cursor: pointer;
    opacity: 0;
    margin: auto 0;
    float: right;
}