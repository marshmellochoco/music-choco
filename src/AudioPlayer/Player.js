import { Slider } from "./Slider";
import './Player.css';
import { useState, useEffect, useCallback } from "react";

export const Player = () => {
    const {
        curTime,
        duration,
        playing,
        volume,
        ended,
        setPlaying,
        setClickedTime,
        setVolume,
        setEnded
    } = AudioPlayer();

    var [previous, setPrevious] = useState([]);
    var [lastVolume, setLastVolume] = useState(volume);

    var [queue, setQueue] = useState([
        {
            icon: 'https://i.scdn.co/image/ab67616d00004851bd5899a4711f4604447166c3',
            title: 'TITLE1',
            artist: '1',
            song: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3'
        },
        {
            icon: 'https://i.scdn.co/image/ab67616d00004851b14cc7a53002e8df68f9a18c',
            title: 'TITLE2',
            artist: '2',
            song: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-2.mp3'
        },
        {
            icon: 'https://cdn.datta.store/auxapi/files/image/cute-cat-kitty-head-kawaii.png',
            title: 'TITLE3',
            artist: '3',
            song: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-3.mp3'
        },
    ]);

    const nextSong = useCallback(
        () => {
        if (queue.length <= 1)
            return;
        setPrevious([queue[0], ...previous]);
        setQueue(queue.slice(1,queue.length));
        setEnded(false);
        }, [previous, queue, setEnded]
    );

    const prevSong = () => {
        if (previous.length < 1) {
            setClickedTime(0);
            return;
        }
        setQueue([previous[0], ...queue]);
        setPrevious(previous.slice(1,previous.length));
    };

    const handleKeyDown = (e) => {
        if (e.shiftKey){
            switch(e.key){
                case 'N': 
                    nextSong(); 
                    break;
                case 'P': 
                    prevSong(); 
                    break;
                default: 
                    return;
            }
        }
        switch(e.key){
            case 'ArrowLeft':
                setClickedTime(curTime - 5);
                break;
            case 'ArrowRight':
                setClickedTime(curTime + 5);
                break;
            case 'ArrowDown':
                setVolume(volume > 0.1 ? volume - 0.1 : 0);
                break;
            case 'ArrowUp':
                setVolume(volume < 0.9 ? volume + 0.1 : 1);
                break;
            case ' ':
                setPlaying(!playing);
                break;
            default:
                return;
        }
    };

    useEffect(() => {
        if (ended === true){
            nextSong()
            setPlaying(false)
            setPlaying(true)
        }
    }, [curTime, ended, setPlaying, nextSong]);

    const formatTime = (seconds, align) => {
        var min = parseInt(Math.floor(seconds/60),10) || 0;
        var sec = Math.floor(seconds%60);
        return (
            <div className="player-time" style={{textAlign: align}}>{(min < 10 ? '0' : '')+ min} : {(sec < 10 ? '0' : '')+ sec}</div>
        )
    };

    return (
        <div className='player' onKeyDown={(e) => handleKeyDown(e)} tabIndex="0">
            <div>
                <audio id='audio-player' src={queue[0].song}/>
                <div className='player-container'>
                    <div className="player-playing">
                        <div className="icon"><img src={queue[0].icon} alt=""/></div>
                        <div className="title">{queue[0].title}</div>
                        <div className="artist">{queue[0].artist}</div>
                    </div>
                    
                    <div className="player-control">
                        <svg xmlns="http://www.w3.org/2000/svg" className="control-button" fill="none" viewBox="0 0 24 24" stroke="currentColor" onClick={() => prevSong()}>
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M11 19l-7-7 7-7m8 14l-7-7 7-7" />
                        </svg>
                        {playing ?
                            <svg onClick={() => setPlaying(false)} xmlns="http://www.w3.org/2000/svg" className="control-button play" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M10 9v6m4-6v6m7-3a9 9 0 11-18 0 9 9 0 0118 0z" />
                            </svg>:
                            <svg onClick={() => setPlaying(true)} xmlns="http://www.w3.org/2000/svg" className="control-button play" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M14.752 11.168l-3.197-2.132A1 1 0 0010 9.87v4.263a1 1 0 001.555.832l3.197-2.132a1 1 0 000-1.664z" />
                                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z" />
                            </svg>
                        }
                        <svg xmlns="http://www.w3.org/2000/svg" className="control-button" fill="none" viewBox="0 0 24 24" stroke="currentColor" onClick={() => nextSong()}>
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M13 5l7 7-7 7M5 5l7 7-7 7" />
                        </svg>
                    </div>
                    
                    <div className="player-volume">
                        {volume === 0 ? 
                        <svg className="volume-icon" onClick={() => {setVolume(lastVolume); }} fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
                            <path fillRule="evenodd" d="M9.383 3.076A1 1 0 0110 4v12a1 1 0 01-1.707.707L4.586 13H2a1 1 0 01-1-1V8a1 1 0 011-1h2.586l3.707-3.707a1 1 0 011.09-.217zM12.293 7.293a1 1 0 011.414 0L15 8.586l1.293-1.293a1 1 0 111.414 1.414L16.414 10l1.293 1.293a1 1 0 01-1.414 1.414L15 11.414l-1.293 1.293a1 1 0 01-1.414-1.414L13.586 10l-1.293-1.293a1 1 0 010-1.414z" clipRule="evenodd" />
                        </svg>:
                        <svg className="volume-icon" onClick={() => {setLastVolume(volume); setVolume(0);}} fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
                            <path fillRule="evenodd" d="M9.383 3.076A1 1 0 0110 4v12a1 1 0 01-1.707.707L4.586 13H2a1 1 0 01-1-1V8a1 1 0 011-1h2.586l3.707-3.707a1 1 0 011.09-.217zM14.657 2.929a1 1 0 011.414 0A9.972 9.972 0 0119 10a9.972 9.972 0 01-2.929 7.071 1 1 0 01-1.414-1.414A7.971 7.971 0 0017 10c0-2.21-.894-4.208-2.343-5.657a1 1 0 010-1.414zm-2.829 2.828a1 1 0 011.415 0A5.983 5.983 0 0115 10a5.984 5.984 0 01-1.757 4.243 1 1 0 01-1.415-1.415A3.984 3.984 0 0013 10a3.983 3.983 0 00-1.172-2.828 1 1 0 010-1.415z" clipRule="evenodd" />
                        </svg>
                        }
                        <Slider onChange={(e) => setVolume(e.target.value/100)} transform={volume*100-100}/>
                    </div>
                    
                    <div className="player-timeline">
                        {formatTime(curTime, "right")}
                        <Slider onChange={(e) => setClickedTime(e.target.value/100*duration)} transform={((curTime / duration) * 100)-100}/>
                        {formatTime(duration, "left")}
                    </div>
                </div>
            </div>
        </div>
    )
}

const AudioPlayer = () => {
    const [duration, setDuration] = useState()
    const [curTime, setCurTime] = useState()
    const [volume, setVolume] = useState()
    const [clickedTime, setClickedTime] = useState()
    const [playing, setPlaying] = useState(false)
    const [ended, setEnded] = useState(false)

    useEffect(() => {
        const audio = document.getElementById('audio-player')

        const setAudioData = () => {
            setDuration(audio.duration)
            setCurTime(audio.currentTime)
            setVolume(audio.volume)
        }

        const setAudioTime = () => setCurTime(audio.currentTime)
        const setAudioVolume = () => setVolume(audio.volume)
        const setAudioEnded = () => setEnded(curTime === duration && playing)

        audio.addEventListener("loadeddata", setAudioData)
        audio.addEventListener("timeupdate", setAudioTime)
        audio.addEventListener("volumechange", setAudioVolume)
        audio.addEventListener("ended", setAudioEnded())

        playing ? audio.play() : audio.pause()

        if (clickedTime != null && clickedTime !== curTime) {
            audio.currentTime = clickedTime
            setClickedTime(null)
        }

        if (volume && volume !== audio.volume)
            audio.volume = volume 

        return () => {
            audio.removeEventListener("loadeddata", setAudioData)
            audio.removeEventListener("timeupdate", setAudioTime)
            audio.removeEventListener("volumechange", setAudioVolume)
            audio.removeEventListener("ended", setAudioEnded)
        }
    }, [playing, clickedTime, curTime, volume, duration])

    return {
        curTime,
        duration,
        playing,
        volume,
        ended,
        setPlaying,
        setClickedTime,
        setVolume,
        setEnded
    }
}