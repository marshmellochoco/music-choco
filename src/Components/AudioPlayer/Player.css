.player{
    font-size: 0.8em;
    width: 98%;
    padding: 10px 1%;
    background-color: #303030;
    color: #d0d0d0;
    overflow: hidden;
    position: absolute;
    bottom: 0;
    height: 8vh;
}

.player:focus{
    outline: none;
}

.player-container {
    display: grid;
    grid-template-columns: 1fr 2fr 1fr;
    grid-template-rows: 2fr 1fr;
}


/* ---------- Playing ---------- */
.player-playing{
    margin: auto 0;
    grid-row: 1/3;
    text-align: left;
    display: grid;
    grid-template-columns: 64px 1fr;
    grid-template-rows: 1fr 1fr;
}

.player-playing .icon{
    grid-row: 1/3;
    width: 64px;
    height: 64px;
}

.player-playing .icon img{
    max-width: 64px;
    max-height: 64px;
    background-color: white;
}

.player-playing .title{
    font-size: 1.2em;
    margin: 0 10px;
    display: flex;
    align-items: flex-end;
    white-space: nowrap;
    overflow: hidden !important;
    text-overflow: ellipsis;
}

.player-playing .artist{
    font-size: 1em;
    margin: 0 10px;
    display: flex;
    align-items: flex-start;
    white-space: nowrap;
    overflow: hidden !important;
    text-overflow: ellipsis;
}

/* ---------- Control ---------- */
.player-control{
    text-align: center;
}

.player-control .control-button{
    height: 100%;
    margin: auto;
    width: 1.8em;
    padding: 0px 12px;
    cursor: pointer;
}

.player-control .play{
    width: 2.5em;
}

.player-control .control-button:hover{
    color: grey;
}


/* ---------- Timeline ---------- */
.player-timeline{
    margin: auto 0;
    text-align: center;
    display: grid;
    grid-template-columns: 1fr 5fr 1fr;
}

.player-timeline .progress{
    overflow: hidden;
    margin: auto 0;
    background-color: grey;
    height: 0.3em;
    border-radius: 2px;
    cursor: pointer;
}

.player-timeline .player-time{
    font-size: 0.9em;
}

/* ---------- Volume ---------- */
.player-volume{
    margin: auto 0 auto auto;
    display: flex;
    width: 40%;
    grid-row: 1/3;
    grid-column: 3;
}

.player-volume .volume-icon{
    border-radius: 50%;
    padding: 4px;
    width: 1.8em;
    cursor: pointer;
}

.player-volume .volume-icon:hover{
    color: grey;
}