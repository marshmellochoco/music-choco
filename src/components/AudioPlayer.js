import { useState, useLayoutEffect } from "react";
import { useDispatch } from "react-redux";
import { useSelector } from "react-redux";
import AudioPlayerComponent from "./AudioPlayerComponent";
import MediaSession from "./MediaSession";
import { setPlaying, setPlayingTrack } from "../store/player/playerAction";

const AudioPlayer = ({ openQueue, setOpenQueue, showPlayer }) => {
    const seekTime = 5;
    var ref = "";
    const dispatch = useDispatch();
    const { playingTrack, playing, loop } = useSelector(
        (state) => state.playerReducer
    );
    const queue = useSelector((state) => state.queueReducer);
    const [time, setTime] = useState(0);
    const [media, setMedia] = useState(playingTrack);

    useLayoutEffect(() => {
        if (playing) ref.current.play();
        else ref.current.pause();
    }, [playing, ref]);

    const setRef = (input) => {
        ref = input;
    };

    const onPlay = () => {
        if (!ref.current) return;
        ref.current.play();
    };

    const onPause = () => {
        if (!ref.current) return;
        ref.current.pause();
    };

    const onPreviousTrack = () => {
        const index = queue.indexOf(playingTrack) - 1;
        if (index >= 0) {
            dispatch(setPlayingTrack(queue[index]));
            dispatch(setPlaying(true));
        } else {
            ref.current.currentTime = 0;
            setTime(0);
        }
    };

    const onNextTrack = () => {
        const index = queue.indexOf(playingTrack) + 1;
        if (index >= queue.length) {
            if (!loop) {
                dispatch(setPlaying(false));
                ref.current.currentTime = 0;
                setTime(100);
            } else {
                dispatch(setPlayingTrack(queue[0]));
                dispatch(setPlaying(true));
            }
        } else {
            dispatch(setPlayingTrack(queue[index]));
            dispatch(setPlaying(true));
        }
    };

    const seekByTime = (t) => {
        if (!ref.current) return;
        ref.current.currentTime += t;
    };

    const seekPrecent = (progress) => {
        if (!ref.current) return;
        ref.current.currentTime =
            (Math.ceil(ref.current.duration) - 1) * progress;
    };

    useLayoutEffect(() => {
        setMedia({
            ...playingTrack,
        });
    }, [playingTrack]);

    return (
        <>
            {media.artists && (
                <MediaSession
                    title={media.title}
                    artist={media.artists.map((a) => a.name).join(", ")}
                    album={media.album.name}
                    artwork={[
                        {
                            src: media.album.image,
                        },
                    ]}
                    onPlay={onPlay}
                    onPause={onPause}
                    onNextTrack={onNextTrack}
                    onPreviousTrack={onPreviousTrack}
                    onSeekForward={() => seekByTime(seekTime)}
                    onSeekBackward={() => seekByTime(-seekTime)}
                />
            )}

            <AudioPlayerComponent
                openQueue={openQueue}
                setOpenQueue={setOpenQueue}
                time={time}
                setRef={setRef}
                onPlay={onPlay}
                onPause={onPause}
                onPreviousTrack={onPreviousTrack}
                onNextTrack={onNextTrack}
                seekPrecent={seekPrecent}
                showPlayer={showPlayer}
            />
        </>
    );
};

export default AudioPlayer;
