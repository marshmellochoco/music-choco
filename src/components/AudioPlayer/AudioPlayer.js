import { useState, useLayoutEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import {} from "react-redux";
import AudioPlayerContainer from "./AudioPlayerContainer";
import MediaSession from "./MediaSession";
import { setPlaying, setPlayingTrack } from "../../store/player/playerAction";

const AudioPlayer = ({ openQueue, setOpenQueue, showPlayer }) => {
    const seekTime = 5;
    const ref = useRef();
    const dispatch = useDispatch();
    const { playingTrack, playing, loop } = useSelector(
        (state) => state.playerReducer
    );
    const queue = useSelector((state) => state.queueReducer);
    const [time, setTime] = useState(0);
    const [media, setMedia] = useState(playingTrack);

    const onPlay = () => {
        if (!ref) return;
        let playPromise = ref.current.play();
        playPromise.catch((err) => {});
        dispatch(setPlaying(true));
    };

    const onPause = () => {
        if (!ref) return;
        ref.current.pause();
        dispatch(setPlaying(false));
    };

    useLayoutEffect(() => {
        if (playing) onPlay();
        else onPause();
        // eslint-disable-next-line
    }, [playing]);

    const onPreviousTrack = () => {
        if (!ref) return;
        const index = queue.indexOf(playingTrack) - 1;
        if (index >= 0) {
            dispatch(setPlayingTrack(queue[index]));
            dispatch(setPlaying(true));
        } else {
            ref.current.currentTime = 0;
            setTime(0);
        }
    };

    const onNextTrack = () => {
        if (!ref) return;
        const index = queue.indexOf(playingTrack);
        if (index >= queue.length - 1) {
            if (loop) {
                dispatch(setPlayingTrack(queue[0]));
                onPlay();
            } else {
                onPause();
                ref.current.currentTime = 0;
                setTime(0);
            }
        } else {
            dispatch(setPlayingTrack(queue[index + 1]));
            onPlay();
        }
    };

    const seekByTime = (t) => {
        if (!ref) return;
        ref.current.currentTime = ref.current.currentTime + t;
    };

    const seekPercent = (progress) => {
        if (!ref) return;
        ref.current.currentTime =
            (Math.ceil(ref.current.duration) - 1) * progress;
    };

    useLayoutEffect(() => {
        setMedia({
            ...playingTrack,
        });
    }, [playingTrack]);

    return (
        <>
            {media.artists && (
                <MediaSession
                    title={media.name}
                    artist={media.artists.map((a) => a.name).join(", ")}
                    album={media.album.name}
                    artwork={[
                        {
                            src: media.album.image,
                        },
                    ]}
                    onPlay={onPlay}
                    onPause={onPause}
                    onNextTrack={onNextTrack}
                    onPreviousTrack={onPreviousTrack}
                    onSeekForward={() => seekByTime(seekTime)}
                    onSeekBackward={() => seekByTime(-seekTime)}
                />
            )}

            <AudioPlayerContainer
                openQueue={openQueue}
                setOpenQueue={setOpenQueue}
                time={time}
                forwardRef={ref}
                onPlay={onPlay}
                onPause={onPause}
                onPreviousTrack={onPreviousTrack}
                onNextTrack={onNextTrack}
                seekPercent={seekPercent}
                showPlayer={showPlayer}
            />
        </>
    );
};

export default AudioPlayer;
