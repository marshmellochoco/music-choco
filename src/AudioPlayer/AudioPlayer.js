import AudioTimeline from './PlayerTimeline';
import VolumeSlider from './VolumeSlider';
import AudioControl from './PlayerControl';
import './AudioPlayer.css';
import AudioPlaying from './NowPlaying';
import { useState, useEffect } from "react";

export default function Audio() {
    const {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    } = AudioPlayer();

    const song1 = {
        icon: 'https://i.scdn.co/image/ab67616d00004851bd5899a4711f4604447166c3',
        title: 'TITLE',
        artist: 'ARTIST',
        song: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3'
    };


    function handleKeyDown(key){
        switch(key){
            case 'ArrowLeft':
                setClickedTime(curTime - 5);
                break;
            case 'ArrowRight':
                setClickedTime(curTime + 5);
                break;
            case 'ArrowDown':
                setVolume(volume > 0.1 ? volume - 0.1 : 0);
                break;
            case 'ArrowUp':
                setVolume(volume < 0.9 ? volume + 0.1 : 1);
                break;
            case ' ':
                setPlaying(!playing);
                break;
            default:
                return
        }
    }

    return (
        <div className='player' onKeyDown={(e) => handleKeyDown(e.key)}>
            <audio id='audio-player' src={song1.song}/>
            <div className='player-container'>
                <AudioPlaying icon={song1.icon} title={song1.title} artist={song1.artist}/>
                <AudioControl playing={playing} setPlaying={setPlaying}/>
                <VolumeSlider volume={volume} setVolume={(volume) => setVolume(volume)}/>
                <AudioTimeline duration={duration} curTime={curTime}/>
            </div>
        </div>
    )
}

function AudioPlayer() {
    const [duration, setDuration] = useState()
    const [curTime, setCurTime] = useState()
    const [volume, setVolume] = useState()
    const [clickedTime, setClickedTime] = useState()
    const [playing, setPlaying] = useState(false)

    useEffect(() => {
        const audio = document.getElementById('audio-player')

        const setAudioData = () => {
            setDuration(audio.duration)
            setCurTime(audio.currentTime)
            setVolume(audio.volume)
        }

        const setAudioTime = () => setCurTime(audio.currentTime)
        const setAudioVolume = () => setVolume(audio.volume)

        audio.addEventListener("loadeddata", setAudioData)
        audio.addEventListener("timeupdate", setAudioTime)
        audio.addEventListener("volumechange", setAudioVolume)

        playing ? audio.play() : audio.pause()

        if (clickedTime && clickedTime !== curTime) {
            audio.currentTime = clickedTime
            setClickedTime(null);
        }

        if (volume && volume !== audio.volume)
            audio.volume = volume; 

        return () => {
            audio.removeEventListener("loadeddata", setAudioData)
            audio.removeEventListener("timeupdate", setAudioTime)
            audio.removeEventListener("volumechange", setAudioVolume)
        }
    }, [playing, clickedTime, curTime, volume])

    return {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    }
}