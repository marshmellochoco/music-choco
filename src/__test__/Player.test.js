import { createStore } from "redux";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { Player } from "../Components/AudioPlayer/Player";
import { rootReducer } from "../Reducers/rootReducer";

const store = createStore(rootReducer);
const testSongData = {
    songId: "60b67f66f747a945dcc5d27c",
    title: "To the Moon - Main Theme",
    duration: 296,
    albumname: "To the Moon (Original Soundtrack)",
    albumId: "60b67f1ff747a945dcc5d27b",
    artist: "Kan Gao",
};
store.dispatch({ type: "SET_SONG_DATA", songData: testSongData });

describe("Player test", () => {
    const tree = renderer.create(
        <Provider store={store}>
            <Player apiUrl={process.env.REACT_APP_API_URL} />
        </Provider>
    );

    it("should render the player correctly", () => {
        expect(tree).toMatchSnapshot();
    });
});
