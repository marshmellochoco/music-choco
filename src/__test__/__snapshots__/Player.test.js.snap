// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player test should render the player correctly 1`] = `
<div
  className="player"
  tabIndex="0"
>
  <div>
    <div
      style={
        Object {
          "height": "0",
          "width": "0",
        }
      }
    />
    <div
      className="player-container"
    >
      <div
        className="player-playing"
      >
        <div
          className="icon"
        >
          <img
            alt=""
            onError={[Function]}
            src="http://118.100.207.12:4000/album/60b67f1ff747a945dcc5d27b/ico"
          />
        </div>
        <div
          className="title"
        >
          To the Moon - Main Theme
        </div>
        <div
          className="artist"
        >
          Kan Gao
        </div>
      </div>
      <div
        className="player-control"
      >
        <svg
          className="control-button prev"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M6,18V6H8V18H6M9.5,12L18,6V18L9.5,12Z"
            fill="currentColor"
          />
        </svg>
        <svg
          className="control-button play"
          fill="none"
          onClick={[Function]}
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.752 11.168l-3.197-2.132A1 1 0 0010 9.87v4.263a1 1 0 001.555.832l3.197-2.132a1 1 0 000-1.664z"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
          />
          <path
            d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
          />
        </svg>
        <svg
          className="control-button next"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M16,18H18V6H16M6,18L14.5,12L6,6V18Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        className="player-volume"
      >
        <svg
          className="volume-icon"
          fill="currentColor"
          onClick={[Function]}
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.383 3.076A1 1 0 0110 4v12a1 1 0 01-1.707.707L4.586 13H2a1 1 0 01-1-1V8a1 1 0 011-1h2.586l3.707-3.707a1 1 0 011.09-.217zM14.657 2.929a1 1 0 011.414 0A9.972 9.972 0 0119 10a9.972 9.972 0 01-2.929 7.071 1 1 0 01-1.414-1.414A7.971 7.971 0 0017 10c0-2.21-.894-4.208-2.343-5.657a1 1 0 010-1.414zm-2.829 2.828a1 1 0 011.415 0A5.983 5.983 0 0115 10a5.984 5.984 0 01-1.757 4.243 1 1 0 01-1.415-1.415A3.984 3.984 0 0013 10a3.983 3.983 0 00-1.172-2.828 1 1 0 010-1.415z"
            fillRule="evenodd"
          />
        </svg>
        <div
          className="slider"
        >
          <div
            className="slider-front"
            style={
              Object {
                "transform": "translateX(0%)",
              }
            }
          />
          <input
            className="slider-back"
            onChange={[Function]}
            type="range"
          />
        </div>
      </div>
      <div
        className="player-timeline"
      >
        <div
          className="player-time"
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          00
           
          :
           
          00
        </div>
        <div
          className="slider"
        >
          <div
            className="slider-front"
            style={
              Object {
                "transform": "translateX(-100%)",
              }
            }
          />
          <input
            className="slider-back"
            onChange={[Function]}
            type="range"
          />
        </div>
        <div
          className="player-time"
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          04
           
          :
           
          56
        </div>
      </div>
    </div>
  </div>
</div>
`;
